### ISSUE TITLE
Unsure if I'm doing anything wrong not importing correctly or using location

### ISSUE THREAD

**Issue by NEprod** [jmathai/elodie#214]
I installed with the instructions from git which everything seemed to install fine. My config file consists of this information.

[MapQuest]
key=5VPxkcLphAAnrNMEOrf2fYPvRO3AxBWz

[Directory]
location=%city, %state
month=%m
year=%Y
date=%year-%month
full_path=%date/%location
# -> 2015-12/Sunnyvale, California

Now when running the import task I get the following it doesn't create the directory structure as requested in config either.

./elodie.py import --destination="/Users/dalesmacbookpro/Google Drive/Google Photos" /Users/dalesmacbookpro/Google\ Drive/Organise/
/Users/dalesmacbookpro/Google Drive/Organise/IMG_1655.JPG -> /Users/dalesmacbookpro/Google Drive/Google Photos/2016-05-May/Unknown Location/2016-05-26_11-39-20-img_1655.jpg
/Users/dalesmacbookpro/Google Drive/Organise/IMG_0221 (1).JPG -> /Users/dalesmacbookpro/Google Drive/Google Photos/2016-08-Aug/Unknown Location/2016-08-20_10-26-01-img_0221 (1).jpg
/Users/dalesmacbookpro/Google Drive/Organise/IMG_1291 (1).MOV -> /Users/dalesmacbookpro/Google Drive/Google Photos/2016-05-May/Unknown Location/2016-05-01_12-54-39-img_1291 (1).mov
/Users/dalesmacbookpro/Google Drive/Organise/IMG_1737 (1).JPG -> /Users/dalesmacbookpro/Google Drive/Google Photos/2017-04-Apr/Unknown Location/2017-04-04_17-54-21-img_1737 (1).jpg
Traceback (most recent call last):
  File "./elodie.py", line 334, in <module>
    main()
  File "/Library/Python/2.7/site-packages/click/core.py", line 716, in __call__
    return self.main(*args, **kwargs)
  File "/Library/Python/2.7/site-packages/click/core.py", line 696, in main
    rv = self.invoke(ctx)
  File "/Library/Python/2.7/site-packages/click/core.py", line 1060, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Library/Python/2.7/site-packages/click/core.py", line 889, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Library/Python/2.7/site-packages/click/core.py", line 534, in invoke
    return callback(*args, **kwargs)
  File "./elodie.py", line 112, in _import
    trash, allow_duplicates)
  File "./elodie.py", line 64, in import_file
    media, allowDuplicate=allow_duplicates, move=False)
  File "/Users/dalesmacbookpro/elodie/elodie/filesystem.py", line 294, in process_file
    directory_name = self.get_folder_path(metadata)
  File "/Users/dalesmacbookpro/elodie/elodie/filesystem.py", line 203, in get_folder_path
    place_name,
  File "/Users/dalesmacbookpro/elodie/elodie/filesystem.py", line 274, in parse_mask_for_location
    folder_name = place_name['default']
KeyError: 'default'

**Comment by Prosatanos** [comment]
I got this same error.

> $  ./elodie.py import --destination "/d/Stage2" /d/Stage1
D:/Stage1\Blake's 4th Birthday (2011)\2011-08-07 14.10.20.jpg -> D:\Stage2\2011-08-Aug\Blake's 4th Birthday (2011)\2011-08-07_06-10-20-2011-08-07 14.10.20.jpg
D:/Stage1\Blake's 4th Birthday (2011)\2011-08-07 14.10.41.jpg -> D:\Stage2\2011-08-Aug\Blake's 4th Birthday (2011)\2011-08-07_06-10-41-2011-08-07 14.10.41.jpg
D:/Stage1\Blake's 4th Birthday (2011)\2011-08-07 14.11.07.jpg -> D:\Stage2\2011-08-Aug\Unknown Location\2011-08-07_06-11-07-2011-08-07 14.11.07.jpg
Traceback (most recent call last):
  File "./elodie.py", line 334, in <module>
    main()
  File "C:\Python27\lib\site-packages\click\core.py", line 716, in __call__
    return self.main(*args, **kwargs)
  File "C:\Python27\lib\site-packages\click\core.py", line 696, in main
    rv = self.invoke(ctx)
  File "C:\Python27\lib\site-packages\click\core.py", line 1060, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "C:\Python27\lib\site-packages\click\core.py", line 889, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "C:\Python27\lib\site-packages\click\core.py", line 534, in invoke
    return callback(*args, **kwargs)
  File "./elodie.py", line 112, in _import
    trash, allow_duplicates)
  File "./elodie.py", line 64, in import_file
    media, allowDuplicate=allow_duplicates, move=False)
  File "C:\Users\user\elodie\elodie\filesystem.py", line 341, in process_file
    directory_name = self.get_folder_path(metadata)
  File "C:\Users\user\elodie\elodie\filesystem.py", line 252, in get_folder_path
    place_name,
  File "C:\Users\user\elodie\elodie\filesystem.py", line 321, in parse_mask_for_location
    folder_name = place_name['default']
KeyError: 'default'

**Comment by NEprod** [comment]
Yea I've tried a full fresh clean install and all I get is the same error message not sure if its something I'm doing or not. Im sure I've got all thats needed installed and more. Would love to try this out so I can have one large collection of organised photos rather than a bundle mess everywhere.

**Comment by jmathai** [comment]
Do you have a file at `~/.elodie/location.json`? Can you try deleting that file and running the same command again?

**Comment by jmathai** [comment]
If that doesn't fix the problem can you try to set `debug = True` in `elodie/constants.py` and paste the output?

**Comment by NEprod** [comment]
Didn't have any file changed the debug to true like you said below is what happened next

./elodie.py import --destination="/Users/dalethomas/Google Drive/Google Photos/" /Users/dalethomas/Google\ Drive/Google\ Photos/
/Users/dalethomas/Google Drive/Google Photos/Paris 2009/PC180059.JPG already exists at /Users/dalethomas/Google Drive/Google Photos/2009-12-Dec/Unknown Location/2009-12-18_12-24-42-pc180059.jpg. Skipping...
/Users/dalethomas/Google Drive/Google Photos/Paris 2009/PC170001.JPG already exists at /Users/dalethomas/Google Drive/Google Photos/2009-12-Dec/Unknown Location/2009-12-17_13-52-50-pc170001.jpg. Skipping...
/Users/dalethomas/Google Drive/Google Photos/Lanzarote 2008/Lanzarote 2008 105.JPG already exists at /Users/dalethomas/Google Drive/Google Photos/2008-07-Jul/Unknown Location/2008-07-20_22-32-17-lanzarote 2008 105.jpg. Skipping...
Traceback (most recent call last):
  File "./elodie.py", line 334, in <module>
    main()
  File "/Library/Python/2.7/site-packages/click/core.py", line 716, in __call__
    return self.main(*args, **kwargs)
  File "/Library/Python/2.7/site-packages/click/core.py", line 696, in main
    rv = self.invoke(ctx)
  File "/Library/Python/2.7/site-packages/click/core.py", line 1060, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Library/Python/2.7/site-packages/click/core.py", line 889, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Library/Python/2.7/site-packages/click/core.py", line 534, in invoke
    return callback(*args, **kwargs)
  File "./elodie.py", line 112, in _import
    trash, allow_duplicates)
  File "./elodie.py", line 64, in import_file
    media, allowDuplicate=allow_duplicates, move=False)
  File "/Users/dalethomas/elodie/elodie/filesystem.py", line 341, in process_file
    directory_name = self.get_folder_path(metadata)
  File "/Users/dalethomas/elodie/elodie/filesystem.py", line 252, in get_folder_path
    place_name,
  File "/Users/dalethomas/elodie/elodie/filesystem.py", line 321, in parse_mask_for_location
    folder_name = place_name['default']
KeyError: 'default'

I have quite a few pictures there but only scans these few then error

**Comment by jmathai** [comment]
Do you happen to know which picture it gets an error on? If you can email it to me at jaisen (at) jmathai (dot) com I can look into it.

**Comment by NEprod** [comment]
I don't know as it doesn't seem to go through one folder as it's jumps from on to another. If if I set it of on just one folder I get the same. It only manages to work through 1-3 photos before it stops none with location. All do have a location data within them as well.

**Comment by jmathai** [comment]
Can you add this print to line 340 of `filesystem.py` and let me know the file name and email the file to me?

```
340        print(_file)
341        directory_name = self.get_folder_path(metadata)
```

**Comment by Prosatanos** [comment]
Done. Ive sent you the file.

**Comment by jmathai** [comment]
I'm unable to reproduce the problem. Can you let me know any additional details like python version, OS, etc?

**Comment by Prosatanos** [comment]
Clean install on Windows 7 x64, Python 2.7.13.

**Comment by NEprod** [comment]
Clean Install on Mac OS seirra Python 2.7.13

**Comment by jmathai** [comment]
Thanks. Can one of you put together a folder of photos that cause this error and zip them up in an email to me? Seems like it errors after 5 or so photos...

I'm not convinced that will help me reproduce the error but it definitely seems like it should be reproducible.

**Comment by Prosatanos** [comment]
Sent!

**Comment by jmathai** [comment]
I created an Ubuntu VM and did a clean install of Elodie but the zip file @Prosatanos sent imported without issues.

```
./elodie.py import --destination=~/test /vagrant/Blake\'s\ 4th\ Birthday\ \(2011\)/
/vagrant/Blake's 4th Birthday (2011)/2011-08-07 14.10.41.jpg -> /home/vagrant/test/2011-08-Aug/Blake's 4th Birthday (2011)/2011-08-07_14-10-41-2011-08-07 14.10.41.jpg
/vagrant/Blake's 4th Birthday (2011)/2011-08-07 15.20.54.jpg -> /home/vagrant/test/2011-08-Aug/Blake's 4th Birthday (2011)/2011-08-07_15-20-54-2011-08-07 15.20.54.jpg
/vagrant/Blake's 4th Birthday (2011)/2011-08-07 15.20.56.jpg -> /home/vagrant/test/2011-08-Aug/Perth/2011-08-07_15-20-56-2011-08-07 15.20.56.jpg
/vagrant/Blake's 4th Birthday (2011)/2011-08-07 14.10.59.jpg -> /home/vagrant/test/2011-08-Aug/Perth/2011-08-07_14-10-59-2011-08-07 14.10.59.jpg
/vagrant/Blake's 4th Birthday (2011)/2011-08-07 14.11.07.jpg -> /home/vagrant/test/2011-08-Aug/Perth/2011-08-07_14-11-07-2011-08-07 14.11.07.jpg
/vagrant/Blake's 4th Birthday (2011)/2011-08-07 14.10.20.jpg -> /home/vagrant/test/2011-08-Aug/Blake's 4th Birthday (2011)/2011-08-07_14-10-20-2011-08-07 14.10.20.jpg
/vagrant/Blake's 4th Birthday (2011)/2011-08-07 15.20.06.jpg -> /home/vagrant/test/2011-08-Aug/Perth/2011-08-07_15-20-06-2011-08-07 15.20.06.jpg
/vagrant/Blake's 4th Birthday (2011)/2011-08-07 14.10.42.jpg -> /home/vagrant/test/2011-08-Aug/Blake's 4th Birthday (2011)/2011-08-07_14-10-42-2011-08-07 14.10.42.jpg
/vagrant/Blake's 4th Birthday (2011)/2011-08-07 14.11.21.jpg -> /home/vagrant/test/2011-08-Aug/Perth/2011-08-07_14-11-21-2011-08-07 14.11.21.jpg
/vagrant/Blake's 4th Birthday (2011)/2011-08-07 15.19.40.jpg -> /home/vagrant/test/2011-08-Aug/Perth/2011-08-07_15-19-40-2011-08-07 15.19.40.jpg
/vagrant/Blake's 4th Birthday (2011)/2011-08-07 14.11.04.jpg -> /home/vagrant/test/2011-08-Aug/Perth/2011-08-07_14-11-04-2011-08-07 14.11.04.jpg
/vagrant/Blake's 4th Birthday (2011)/2011-08-07 15.20.48.jpg -> /home/vagrant/test/2011-08-Aug/Perth/2011-08-07_15-20-48-2011-08-07 15.20.48.jpg
/vagrant/Blake's 4th Birthday (2011)/2011-08-07 15.20.33.jpg -> /home/vagrant/test/2011-08-Aug/Perth/2011-08-07_15-20-33-2011-08-07 15.20.33.jpg
/vagrant/Blake's 4th Birthday (2011)/2011-08-07 15.19.54-2.jpg -> /home/vagrant/test/2011-08-Aug/Perth/2011-08-07_15-19-54-2011-08-07 15.19.54-2.jpg
****** SUMMARY ******
Metric      Count
--------  -------
Success        14
Error           0
```

**Comment by Prosatanos** [comment]
Try a Windows box.

**Comment by Prosatanos** [comment]
Tried with a clean install on Ubuntu 17.04, but still the same. For comparison, can you please list the commands you entered to install elodie and all required dependencies?

**Comment by jevy** [comment]
Hey y'all! Just did a fresh install of elodie on a near vanilla MBP. Got this on my first try:
``
elodie jevin$ python elodie.py import --destination="/Users/jevin.maltais/Downloads/pics_temp" "/Users/jevin.maltais/Dropbox/Family Pics To Backup"
Traceback (most recent call last):
  File "elodie.py", line 345, in <module>
    main()
  File "/usr/local/lib/python2.7/site-packages/click/core.py", line 716, in __call__
    return self.main(*args, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/click/core.py", line 696, in main
    rv = self.invoke(ctx)
  File "/usr/local/lib/python2.7/site-packages/click/core.py", line 1060, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/local/lib/python2.7/site-packages/click/core.py", line 889, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/lib/python2.7/site-packages/click/core.py", line 534, in invoke
    return callback(*args, **kwargs)
  File "elodie.py", line 115, in _import
    trash, allow_duplicates)
  File "elodie.py", line 64, in import_file
    media, allowDuplicate=allow_duplicates, move=False)
  File "/Users/jevin.maltais/code_github/elodie/elodie/filesystem.py", line 341, in process_file
    directory_name = self.get_folder_path(metadata)
  File "/Users/jevin.maltais/code_github/elodie/elodie/filesystem.py", line 252, in get_folder_path
    place_name,
  File "/Users/jevin.maltais/code_github/elodie/elodie/filesystem.py", line 321, in parse_mask_for_location
    folder_name = place_name['default']
KeyError: 'default'
``

**Comment by dbx77** [comment]
Hi, getting same issue on a Synology DSM. Running without virtualenv. It ran fine for about 500 images but now is getting this error. Tried deleting location.json as well.

**Comment by jmathai** [comment]
@dbx77 does it now fail repeatedly on the same file?

**Comment by dbx77** [comment]
It actually it fails right away. However, I forgot to mention that I used sudo. I was doing that because of some file permissions. But I recall now that last night I called it without sudo. 

I just tried again without sudo and it is going through the files, skipping them since they are already processed. I will check later to see how it goes. 

Not sure why when I call it with sudo fails right away.  Thank you!

**Comment by jmathai** [comment]
@dbx77 if you have the error you received when you ran it via `sudo` that'd be helpful. Let me know if it finishes without `sudo`.

**Comment by jmathai** [comment]
I'm wondering if this PR from @ostavnaas might fix this issue. Can someone try the simple change in this diff and let me know if it helps?

https://github.com/jmathai/elodie/pull/232/files

**Comment by michaeltieso** [comment]
Is this still an issue? It seems to be ignoring the `config.ini` even if I include the key.

**Comment by ostavnaas** [comment]
Have you moved `config.ini` to `~/.elodie/config.ini`?

**Comment by adamklaff** [comment]
Just FYI, I get the same error as well. Tried running with a bunch of photos as a test on my Mac and received the following error:

Traceback (most recent call last):
  File "./elodie.py", line 345, in <module>
    main()
  File "/Library/Python/2.7/site-packages/click/core.py", line 716, in __call__
    return self.main(*args, **kwargs)
  File "/Library/Python/2.7/site-packages/click/core.py", line 696, in main
    rv = self.invoke(ctx)
  File "/Library/Python/2.7/site-packages/click/core.py", line 1060, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Library/Python/2.7/site-packages/click/core.py", line 889, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Library/Python/2.7/site-packages/click/core.py", line 534, in invoke
    return callback(*args, **kwargs)
  File "./elodie.py", line 115, in _import
    trash, allow_duplicates)
  File "./elodie.py", line 64, in import_file
    media, allowDuplicate=allow_duplicates, move=False)
  File "/Users/adam/dev/elodie/elodie/filesystem.py", line 341, in process_file
    directory_name = self.get_folder_path(metadata)
  File "/Users/adam/dev/elodie/elodie/filesystem.py", line 252, in get_folder_path
    place_name,
  File "/Users/adam/dev/elodie/elodie/filesystem.py", line 321, in parse_mask_for_location
    folder_name = place_name['default']
KeyError: 'default'

**Comment by jmathai** [comment]
I'm still trying to reproduce this. Could someone that's getting this error provide me with the following via here or email jaisen(at)jmathai(dot)com? // @adamklaff, @dbx77, @Prosatanos, @NEprod

* A copy of the following files.
    * `~/.elodie/config.ini`
    * `~/.elodie/location.json`

**Comment by rkoster** [comment]
With some additional debugging, it seems the problem is related to invalid caching of a `None` location, the first photo (after I deleted `location.json`) processes fine, however the second photo get an empty `{}` [from the cache](https://github.com/jmathai/elodie/blob/master/elodie/geolocation.py#L132).
```
metadata:
{'album': None, 'extension': u'jpg', 'mime_type': 'image/jpeg', 'date_taken': time.struct_time(tm_year=2016, tm_mon=5, tm_mday=26, tm_hour=16, tm_min=33, tm_sec=59, tm_wday=3, tm_yday=147, tm_isdst=0), 'title': None, 'latitude': 17056881.853375, 'base_name': u'P1070523', 'original_name': None, 'longitude': 17056881.853375, 'directory_path': u'/photo/old/2016-07-09'}
cached_place_name:
None
geolocation_info:
None
place_name:
{'default': 'Unknown Location'}
metadata:
{'album': None, 'extension': u'jpg', 'mime_type': 'image/jpeg', 'date_taken': time.struct_time(tm_year=2015, tm_mon=11, tm_mday=2, tm_hour=10, tm_min=22, tm_sec=53, tm_wday=0, tm_yday=306, tm_isdst=0), 'title': None, 'latitude': 17056881.853375, 'base_name': u'P1020708', 'original_name': None, 'longitude': 17056881.853375, 'directory_path': u'/photo/old/2015-11-11'}
cached_place_name:
{}
place_name:
{}
```

**Comment by rkoster** [comment]
this error was introduced here: https://github.com/jmathai/elodie/commit/6777c3258874bcc001002ca6ca2b52a2760f24f6

**Comment by ethack** [comment]
@jmathai I think I found a way to reproduce this error.  I was getting the exact same error, and I tracked my issue down to a mistyped MapQuest API key.  The MapQuest server responded with a 403 status code and caused [this line](https://github.com/jmathai/elodie/blob/master/elodie/geolocation.py#L185) to trigger an exception and thus the function to return None.  So to reproduce, use any image with geocoordinate info to trigger the MapQuest query but use an invalid API key in your config file.

FWIW, I had merged in pull request #233 at the time I did all this.

**Comment by ostavnaas** [comment]
@ethack I was not able to reproduce this scenario with pull request #233. Are you sure the result was not cached in location.json from a previously run, or not having the latest commit https://github.com/jmathai/elodie/pull/233/commits/996c05fbd402f4cdbf72852b14360685f596764d from that pull request.

I got this message with debug mode
`The AppKey submitted with this request is invalid.
No JSON object could be decoded
`
And all photos was moved to directory named Unknown Location

**Comment by ethack** [comment]
You're right, I can't reproduce the error today with 996c05fbd402f4cdbf72852b14360685f596764d applied. I must have been wrong about the order I made the changes. 

On the plus side, 996c05fbd402f4cdbf72852b14360685f596764d appears to fix the issue, at least for me. If I remove that commit I get the error (with the invalid API key), but if I have the commit in then the image is moved to Unknown Location like you said.

I didn't see the same message you did, but I'm not sure if I enabled debug mode correctly or am looking in the right place either. It at least ran without errors and moved the photo though.
